sha2/src/sha512/wasm32.rs

```diff
@@ -0,0 +1,176 @@
+#![allow(clippy::many_single_char_names)]
+
+use core::arch::wasm32::*;
+use core::mem::size_of;
+
+use crate::consts::K64;
+
+pub fn compress(state: &mut [u64; 8], blocks: &[[u8; 128]]) {
+    let mut ms = [u64x2(0, 0); 8];
+    let mut x = [u64x2(0, 0); 8];
+
+    for block in blocks {
+        unsafe {
+            let mut current_state = *state;
+            load_data(&mut x, &mut ms, block.as_ptr().cast());
+            rounds_0_63(&mut current_state, &mut x, &mut ms);
+            rounds_64_79(&mut current_state, &ms);
+            accumulate_state(state, &current_state);
+        }
+    }
+}
+
+#[inline(always)]
+unsafe fn load_data(x: &mut [v128; 8], ms: &mut MsgSchedule, data: *const v128) {
+    macro_rules! unrolled_iterations {
+        ($($i:literal),*) => {$(
+            x[$i] = v128_load(data.add($i).cast());
+            x[$i] = i8x16_shuffle::<7,6,5,4,3,2,1,0,15,14,13,12,11,10,9,8>(x[$i], x[$i]);
+
+            let y = i64x2_add(
+                x[$i],
+                v128_load(K64.as_ptr().add(2 * $i).cast()),
+            );
+
+            ms[$i] = y;
+        )*};
+    }
+
+    unrolled_iterations!(0, 1, 2, 3, 4, 5, 6, 7);
+}
+
+#[inline(always)]
+unsafe fn rounds_0_63(current_state: &mut State, x: &mut [v128; 8], ms: &mut MsgSchedule) {
+    let mut k64_idx: usize = SHA512_BLOCK_WORDS_NUM;
+
+    for _ in 0..4 {
+        for j in 0..8 {
+            let k64 = v128_load(K64.as_ptr().add(k64_idx).cast());
+            let y = sha512_update_x(x, k64);
+
+            {
+                let ms = cast_ms(ms);
+                sha_round(current_state, ms[2 * j]);
+                sha_round(current_state, ms[2 * j + 1]);
```

Comment:
Maybe it's better to write it like this?
```rust
let ms = ms[j];
sha_round(current_state, u64x2_extract_lane::<0>(ms));
sha_round(current_state, u64x2_extract_lane::<1>(ms));
```
Same for other uses of `cast_ms`.

Reply:
Agreed, I like getting rid of `cast_ms`.